algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/molecular/GeneHasSufficientCopyNumber.kt

```diff
@@ -0,0 +1,106 @@
+package com.hartwig.actin.algo.evaluation.molecular
+
+import com.hartwig.actin.algo.evaluation.EvaluationFactory
+import com.hartwig.actin.datamodel.algo.Evaluation
+import com.hartwig.actin.datamodel.molecular.driver.GeneRole
+import com.hartwig.actin.datamodel.molecular.MolecularTest
+import com.hartwig.actin.datamodel.molecular.driver.ProteinEffect
+import com.hartwig.actin.datamodel.molecular.driver.CopyNumber
+import java.time.LocalDate
+
+private enum class CopyNumberEvaluation {
+    REPORTABLE_SUFFICIENT_COPY_NUMBER,
+    SUFFICIENT_COPY_NUMBER_WITH_LOSS_OF_FUNCTION,
+    SUFFICIENT_COPY_NUMBER_ON_NON_ONCOGENE,
+    UNREPORTABLE_SUFFICIENT_COPY_NUMBER,
+    INSUFFICIENT_COPY_NUMBER;
+
+    companion object {
+        fun fromCopyNumber(copyNumber: CopyNumber, requestedMinCopyNumber: Int): CopyNumberEvaluation {
+            return if (copyNumber.canonicalImpact.minCopies >= requestedMinCopyNumber) {
+                when {
+                    copyNumber.geneRole == GeneRole.TSG -> {
+                        SUFFICIENT_COPY_NUMBER_ON_NON_ONCOGENE
+                    }
+
+                    copyNumber.proteinEffect == ProteinEffect.LOSS_OF_FUNCTION
+                            || copyNumber.proteinEffect == ProteinEffect.LOSS_OF_FUNCTION_PREDICTED -> {
+                        SUFFICIENT_COPY_NUMBER_WITH_LOSS_OF_FUNCTION
+                    }
+
+                    !copyNumber.isReportable -> {
+                        UNREPORTABLE_SUFFICIENT_COPY_NUMBER
+                    }
+
+                    else -> {
+                        REPORTABLE_SUFFICIENT_COPY_NUMBER
+                    }
+                }
+            } else {
+                INSUFFICIENT_COPY_NUMBER
+            }
+        }
+    }
+}
+
+class GeneHasSufficientCopyNumber(private val gene: String, private val requestedMinCopyNumber: Int, maxTestAge: LocalDate? = null) :
+    MolecularEvaluationFunction(maxTestAge) {
+
+    override fun genes() = listOf(gene)
+
+    override fun evaluate(test: MolecularTest): Evaluation {
+        val evaluatedCopyNumbers: Map<CopyNumberEvaluation, Set<String>> = test.drivers.copyNumbers
+            .filter { it.gene == gene }
+            .groupingBy { CopyNumberEvaluation.fromCopyNumber(it, requestedMinCopyNumber) }
+            .fold(emptySet()) { acc, copyNumber -> acc + copyNumber.event }
```

Comment:
Interesting! 🙂 I probably would have done:
```suggestion
            .groupBy { CopyNumberEvaluation.fromCopyNumber(it, requestedMinCopyNumber) }
            .mapValues { it.value.map(CopyNumber::event).toSet() }
```
but it seems like your way might actually be preferable as it lazily produces each set instead of producing an intermediate map: https://kotlinlang.org/docs/collection-grouping.html.